Rollout/recreate visualization for Sidebar panel

sahil143 opened the pull request:
Odc-story: https://jira.coreos.com/browse/ODC-1251

**Recreate**
![recreate-sidebar](https://user-images.githubusercontent.com/9278015/63763075-9f75b900-c8e1-11e9-8769-ec08dfa599c6.gif)
**Rolling**
![rolling-sidebar](https://user-images.githubusercontent.com/9278015/63763076-9f75b900-c8e1-11e9-8063-d7a07e81fc01.gif)

rohitkrai03 commented:
@sahil143 This one has changes from https://github.com/openshift/console/pull/2456/ as well. I have some review comments that you already took care of in the other PR. Maybe you can merge both PRs into one with two different commits or reviewing this PR makes sense after the first one gets merged.

sahil143 commented:
@rohitkrai03 only common changes are from pod-util file. once that PR is merged I will rebase this one and we will be good to go.

sahil143 commented:
/hold

sahil143 commented:
/hold cancel

andrewballantyne commented on frontend/packages/console-shared/src/components/pod/PodRing.tsx:
```diff
@@ -55,6 +55,8 @@ const PodRing: React.FC<PodRingProps> = ({ pods, obj, resourceKind, path, impers
 
   const isKnative = _.get(obj, 'metadata.ownerReferences[0].kind') === 'Revision';
 
+  const resourceObj = rc || obj;
```
Isn't `rc` required? Is it expected to be null or undefined here?

sahil143 replied:
When a DC is created both current and Previous are undefined until an RC is created. In that case `rc` will be undefined. Also, StatefulSets don't have current/previous RC data and if we start to show SS in SideBar `rc` will prop will not be passed. I will make `rc` optional. Thanks for noticing. :)

sahil143 replied:
fixed in https://github.com/openshift/console/pull/2496/commits/8ee400e6ae62c6d9de3fe35a290b834415a1bed8

andrewballantyne commented:
I noticed no arrow between the pod charts... it exists in the docs, so I have posted a comment on there and tagged you (@sahil143) and @serenamarie125 for more information.

andrewballantyne commented:
@sahil143 I don't think we want the scale up/down arrows next to the charts during the rollout. 
![image](https://user-images.githubusercontent.com/8126518/66658549-b5300b00-ec0f-11e9-90a9-1a5963f077cd.png)

They don't do anything and I don't think they would have any purpose during the rollout.

sahil143 commented:
> @sahil143 I don't think we want the scale up/down arrows next to the charts during the rollout.

Agreed, Makes sense. Question is do we disable them or remove them altogether during the rollout?

cc: @serenamarie125 ^^

andrewballantyne commented:
> Agreed, Makes sense. Question is do we disable them or remove them altogether during the rollout?
> 
> cc: @serenamarie125 ^^

Mockups show the arrows removed.

andrewballantyne commented on frontend/packages/dev-console/src/components/topology/shapes/PodSet.tsx:
```diff
@@ -0,0 +1,67 @@
+  const { podStatusOuterRadius, podStatusInnerRadius, podStatusStrokeWidth } = calculateRadius(
+    size,
+  );
+  const { innerPodStatusOuterRadius, innerPodStatusInnerRadius } = calculateInnerPodStatusRadius(
```
Was going to suggest we reuse `calculateRadius` (instead of creating `calculateInnerPodStatusRadius`) and make it accept an optional `strokeWidth` - but that code makes me uncomfortable lol. I think we are best writing tests around it and then worrying about consolidating that logic.

andrewballantyne replied:
I created https://jira.coreos.com/browse/ODC-2012 to track the above change.

andrewballantyne commented on frontend/packages/dev-console/src/components/topology/shapes/WorkloadNode.tsx:
```diff
@@ -33,7 +30,7 @@ const WorkloadNode: React.FC<NodeProps<WorkloadData>> = ({
       y={y}
       outerRadius={radius}
-      innerRadius={radius * 0.55}
+      innerRadius={radius * 0.45}
```
Can we gate the shrinkage of the center image around the existence of the second donut? I don't think we need to shrink the icon always - I'm not sure how often rollout will be in-use... but the standing look is definitely the first impression people will see of the node and a smaller icon isn't helping anything.

sahil143 replied:
fixed in https://github.com/openshift/console/pull/2496/commits/6353ee79b77029607673e09df07a9099b0fb4a9b

sahil143 commented:
@andrewballantyne Added the arrow b/w donuts and removed the scaling buttons during a rollout
fixed in https://github.com/openshift/console/pull/2496/commits/f0e6a9fa4ce2d119b9814613634e409a0897898a
![Screenshot from 2019-10-11 22-00-37](https://user-images.githubusercontent.com/9278015/66668571-2d7be800-ec73-11e9-9cd8-be66ff2773a5.png)

sahil143 commented on frontend/packages/console-shared/src/components/pod/PodRingSet.tsx:
```diff
@@ -0,0 +1,57 @@
+          />
+        </SplitItem>
+        <SplitItem>
+          {inProgressDeploymentData && (
```
fixed in https://github.com/openshift/console/pull/2496/commits/f0e6a9fa4ce2d119b9814613634e409a0897898a

sahil143 commented on frontend/packages/console-shared/src/components/pod/PodRingSet.tsx:
```diff
@@ -0,0 +1,57 @@
+  const progressRC = inProgressDeploymentData && current;
+  const completedRC = !!inProgressDeploymentData && completedDeploymentData ? previous : current;
+  return (
+    <React.Fragment>
```
fixed in https://github.com/openshift/console/pull/2496/commits/f0e6a9fa4ce2d119b9814613634e409a0897898a

sahil143 commented on frontend/packages/console-shared/src/components/pod/PodRingSet.tsx:
```diff
@@ -0,0 +1,57 @@
+      <Split gutter="lg">
+        <SplitItem>
+          <PodRing
+            key={inProgressDeploymentData ? 'deploy' : 'notDeploy'}
```
Yes, whenever the rollout is complete it retriggers the animation on the donut. This key is to kill the animation.

sahil143 commented on frontend/packages/console-shared/src/utils/pod-utils.ts:
```diff
@@ -142,3 +146,35 @@ export const isIdled = (deploymentConfig: K8sResourceKind): boolean => {
+  previous: PodControllerOverviewItem,
+  isRollingOut: boolean,
+): { inProgressDeploymentData: Pod[] | null; completedDeploymentData: Pod[] | null } => {
+  const strategy = _.get(dc, ['spec', 'strategy', 'type']);
```
I used _.get here because for StatefulSets and daemonSets `strategy` doesn't exist. For straight accessing, I would need to apply check for spec then the strategy and then get the type.

andrewballantyne replied:
Fair enough - use the 3rd param (even if you're setting it to `null`) and type `strategy`, we have TypeScript so we should leverage readability through types.

sahil143 replied:
fixed in https://github.com/openshift/console/pull/2496/commits/6353ee79b77029607673e09df07a9099b0fb4a9b

andrewballantyne commented:
> @andrewballantyne Added the arrow b/w donuts and removed the scaling buttons during a rollout
> fixed in [f0e6a9f](https://github.com/openshift/console/commit/f0e6a9fa4ce2d119b9814613634e409a0897898a)
> ![Screenshot from 2019-10-11 22-00-37](https://user-images.githubusercontent.com/9278015/66668571-2d7be800-ec73-11e9-9cd8-be66ff2773a5.png)

This looks a lot better. Very straight forward that the right donut is scaling up to replace the left donut.

sahil143 commented on frontend/packages/console-shared/src/utils/resource-utils.ts:
```diff
@@ -410,7 +432,7 @@ export class TransformResourceData {
     return _.filter(
       ownedRC,
-      (rc) => _.get(rc, 'status.replicas') || getDeploymentConfigVersion(rc) === currentVersion,
+      (rc) => _.get(rc, 'status.replicas') || getDeploymentConfigVersion(rc) >= currentVersion - 1,
```
If the latest version is 4 and rollout is started then the newly create RC would get version 5 which will become current and the previous will become version 4. 
So, With this condition, currentVersion is 5 and I need previous and current. That's why the `-1`.

sahil143 commented on frontend/packages/console-shared/src/utils/pod-utils.ts:
```diff
@@ -142,3 +146,35 @@ export const isIdled = (deploymentConfig: K8sResourceKind): boolean => {
+  current: PodControllerOverviewItem,
+  previous: PodControllerOverviewItem,
+  isRollingOut: boolean,
+): { inProgressDeploymentData: Pod[] | null; completedDeploymentData: Pod[] | null } => {
```
As discussed before, `Pod[]` is a duplicate of `PodKind` with some extra optional properties and Pod is being used in dev console and Podkind is used in the console. fixed the or-type `null` in https://github.com/openshift/console/pull/2496/commits/6353ee79b77029607673e09df07a9099b0fb4a9b

andrewballantyne replied:
I understand that we did talk about it on the refactor PR... but propagating this logic is just going to make it harder to read the code. We can't, as a collective, be referencing historical conversations to explain duplication... otherwise adoption of the code will just grow more complex and difficult to do.

Is there anyway we can covert over to using one or the other (looks like `PodKind` is the official one and `Pod` was a duplicate with smaller # of usages)? Or extending at least so it's traceable. Seeing `current` and `previous` providing a `PodKind[]` that you're just assigning to `Pod[]` like it's the same thing is on the face confusing.

Something like:
```
export interface Pod extends PodKind {
  id?: string;
  name?: string;
}
```
Not sure if `id` and `name` are valid... But at least this way it's more polymorphism and less "typescript deep comparing and not objecting".

sahil143 replied:
Agreed! fixed in https://github.com/openshift/console/pull/2496/commits/e4b602178af804b74f431046aa9706db239aadab

sahil143 commented on frontend/packages/console-shared/src/utils/resource-utils.ts:
```diff
@@ -370,6 +372,26 @@ const getIdledStatus = (
+  current: PodControllerOverviewItem,
+  previous: PodControllerOverviewItem,
+): boolean => {
+  const strategy = _.get(dc, ['spec', 'strategy', 'type']);
```
fixed in https://github.com/openshift/console/pull/2496/commits/6353ee79b77029607673e09df07a9099b0fb4a9b

sahil143 commented on frontend/packages/console-shared/src/utils/resource-utils.ts:
```diff
@@ -370,6 +372,26 @@ const getIdledStatus = (
+  const phase = current && current.phase;
+  const currentRC = current && current.obj;
+  const notFailedOrCancelled =
+    current && phase !== DEPLOYMENT_PHASE.cancelled && phase !== DEPLOYMENT_PHASE.failed;
```
fixed in https://github.com/openshift/console/pull/2496/commits/6353ee79b77029607673e09df07a9099b0fb4a9b

andrewballantyne commented on frontend/packages/console-shared/src/utils/resource-utils.ts:
```diff
@@ -377,11 +377,15 @@ const getRolloutStatus = (
-  const strategy = _.get(dc, ['spec', 'strategy', 'type']);
+  const {
+    spec: {
+      strategy: { type: strategy },
```
I believe a test just made this an error (unless I misread the test failure).

sahil143 replied:
Yes, It did because Mock data doesn't have a strategy property defined in them. updated mock data in https://github.com/openshift/console/pull/2496/commits/3b8977c4a7be891a8440386ebc7ebd9609b646f1

vikram-raj commented on frontend/packages/console-shared/src/components/pod/PodRingSet.tsx:
```diff
@@ -0,0 +1,65 @@
+        <React.Fragment>
+          <SplitItem>
+            <Bullseye>
+              <LongArrowAltRightIcon size="xl" color="#737679" />
```
Please user ```patternfly/react-token``` variable.

```import { global_Color_200 as color200 } from '@patternfly/react-tokens';```

sahil143 replied:
fixed in https://github.com/openshift/console/pull/2496/commits/7c710a2c732716fb8b688fcb97b5f3a19a9472d2

andrewballantyne commented:
LGTM - please rebase and squash @sahil143

andrewballantyne commented:
/lgtm

rohitkrai03 commented:
/lgtm

sahil143 commented:
/assign @christianvogt

christianvogt commented:
/approve
